src/Microsoft.Extensions.WebSockets/Internal/ChannelExtensions.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Buffers;
+using System.Threading;
+using System.Threading.Tasks;
+using Channels;
+
+namespace Microsoft.Extensions.WebSockets.Internal
+{
+    public static class ChannelExtensions
+    {
+        public static ValueTask<ReadableBuffer> ReadAtLeastAsync(this IReadableChannel input, int minimumRequiredBytes) => ReadAtLeastAsync(input, minimumRequiredBytes, CancellationToken.None);
+
+        // TODO: Pull this up to Channels. We should be able to do it there without allocating a Task<T> in any case (rather than here where we can avoid allocation
+        // only if the buffer is already ready and has enough data.
+        public static ValueTask<ReadableBuffer> ReadAtLeastAsync(this IReadableChannel input, int minimumRequiredBytes, CancellationToken cancellationToken)
+        {
+            var awaiter = input.ReadAsync(/* cancellationToken */);
+
+            // Short-cut path!
+            if (awaiter.IsCompleted)
+            {
+                // We have a buffer, is it big enough?
+                var buffer = awaiter.GetResult();
+
+                if (buffer.Length >= minimumRequiredBytes)
+                {
+                    return new ValueTask<ReadableBuffer>(buffer);
+                }
+
+                // Buffer wasn't big enough, mark it as examined and continue to the "slow" path below
+                input.Advance(
+                    consumed: buffer.Start,
+                    examined: buffer.End);
+            }
+            return new ValueTask<ReadableBuffer>(ReadAtLeastSlowAsync(awaiter, input, minimumRequiredBytes, cancellationToken));
+        }
+
+        public static Memory<byte> ToMemory(this ReadableBuffer buffer)
+        {
+            if (buffer.IsSingleSpan)
+            {
+                return buffer.First;
+            }
+            else
+            {
+                var mem = new Memory<byte>(new byte[buffer.Length]);
```

Comment:
on surface, this looks like yet another place where it would be handy to lease from the memory pool; I haven't seen what uses this yet, though - reserving more input!